### DISCUSSION TITLE
VPI based (tiny) RTL Lint checker

### DISCUSSION THREAD

**Discussion by svenka3** [steveicarus/iverilog#758] [Ideas]
Hello,
  I ported a small Lint checker from http://project-veripage.com/vpilint.php

I masked most of what is seemingly unsupported in Icarus as of now. Please take a look at it via:

https://github.com/svenka3/ivl_uvm/tree/main/ivl_uvm_vpi/ivl_vlog_lint

Any comments, welcome!

Srini

**Comment by caryr** [comment]
I would switch the code from using `tf_dofinish()` to `vpi_control()` so you are only using the more modern VPI calls. The tf/acc routines are completely obsolete and have not been listed in the standard for a long time.

**Reply by svenka3** [reply]
Good point - a legacy from old code, will fix it. Thanks

**Comment by caryr** [comment]
Also, you have memory leaks in your code. The result from `strdup()` and `malloc()` need to have `free()` called to release the memory.

The argument iterator is not released correctly when extra arguments are given. From the standard:

```
vpi_release_handle() may be used to free memory created for iterator objects. The iterator object shall
automatically be freed when vpi_scan() returns NULL because it has either completed an object traversal or
encountered an error condition. If neither of these conditions occurs (which can happen if the code breaks
out of an iteration loop before it has scanned every object), vpi_release_handle() should be called to free
any memory allocated for the iterator.
```

**Reply by svenka3** [reply]
Thanks for deeper look. Yes this was pointed out by a colleague earlier today here. Will clean up.

Appreciate your inputs.

Cheers
